New HTB Account

Hi,
Now that HTB is using the new "HTB Account" which uses the details from the Labs platform, do you have plans to use this login system? Currently neither my Academy or Labs password works and I get the following error:
```
panic: runtime error: slice bounds out of range [1:0]

goroutine 1 [running]:
main.getModulePages({0xc0001dc000, 0x295d}, {0xc000014200, 0x34})
        /home/tut/Programming/go/htb-academy-md/src/export.go:217 +0x177
main.getModule({0xc000014200, 0x34}, {{0xc000230780?, 0x17?}, {0xc000108000?, 0x3dd7c8?}})
        /home/tut/Programming/go/htb-academy-md/src/export.go:142 +0x3b4
main.main()
        /home/tut/Programming/go/htb-academy-md/src/main.go:13 +0xd7
```
My HTB Account doesnt have 2FA enabled.

Thanks

--- a/go.mod
+++ b/go.mod
@@ -5,10 +5,12 @@ go 1.20
 require (
 	github.com/JohannesKaufmann/html-to-markdown v1.3.7
 	golang.org/x/net v0.8.0
+	golang.org/x/term v0.6.0
 )
 
 require (
 	github.com/PuerkitoBio/goquery v1.8.1 // indirect
 	github.com/andybalholm/cascadia v1.3.1 // indirect
+	golang.org/x/sys v0.6.0 // indirect
 	gopkg.in/yaml.v2 v2.4.0 // indirect
 )

--- a/src/args.go
+++ b/src/args.go
@@ -4,6 +4,8 @@ import (
 	"flag"
 	"fmt"
 	"os"
+
+	"golang.org/x/term"
 )
 
 type Args struct {
@@ -15,8 +17,8 @@ type Args struct {
 
 func getArguments() Args {
 	var mFlag = flag.String("m", "", "(REQUIRED) Academy Module URL to the first page.")
-	var eFlag = flag.String("e", "", "(REQUIRED) Email for your HTB Academy account.")
-	var pFlag = flag.String("p", "", "(REQUIRED) Password for your HTB Academy account.")
+	var eFlag = flag.String("e", "", "(REQUIRED) Email for your HTB account.")
+	var pFlag = flag.String("p", "", "Password for your HTB account.")
 	var imgFlag = flag.Bool("local_images", false, "Save images locally rather than referencing the URL location.")
 	flag.Parse()
 	arg := Args{
@@ -26,9 +28,20 @@ func getArguments() Args {
 		localImages: *imgFlag,
 	}
 
-	if arg.moduleUrl == "" || arg.email == "" || arg.password == "" {
-		fmt.Println("Missing arguments, please use the -h option to display the arguments required to run this application!")
+	if arg.moduleUrl == "" || arg.email == "" {
+		fmt.Println("Missing required arguments for module URL and HTB email. Please use the -h option for help.")
 		os.Exit(1)
+	}
+
+	if arg.password == "" {
+		fmt.Print("Enter Password: ")
+		passwordBytes, err := term.ReadPassword(int(os.Stdin.Fd()))
+		if err != nil {
+			fmt.Println("\nError reading password:", err)
+			os.Exit(1)
+		}
+		arg.password = string(passwordBytes)
+		fmt.Println()
 	}
 
 	return arg

--- a/src/export.go
+++ b/src/export.go
@@ -1,122 +1,162 @@
 package main
 
 import (
+	"bytes"
+	"encoding/json"
 	"fmt"
 	"golang.org/x/net/html"
+	"golang.org/x/net/publicsuffix"
 	"io"
 	"math/rand"
 	"net/http"
+	"net/http/cookiejar"
 	"net/url"
 	"os"
 	"strings"
 )
 
+type userAgentTransport struct {
+	Transport http.RoundTripper
+	UserAgent string
+}
+
 type Auth struct {
-	loginToken string
-	cookies    string
+	cookies string
+}
+
+type Credentials struct {
+	Email    string `json:"email"`
+	Password string `json:"password"`
+}
+
+type LoginResponse struct {
+	IntendedRoute string `json:"intended_route"`
 }
 
 func authenticate(email, password string) Auth {
-	auth := getLoginTokenAndCookies()
-	payload := "_token=" + auth.loginToken + "&email=" + url.QueryEscape(email) + "&password=" + url.QueryEscape(password)
-
-	//proxy, _ := url.Parse("http://localhost:8080")
-	//tr := &http.Transport{Proxy: http.ProxyURL(proxy), TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}
-	client := &http.Client{
-		CheckRedirect: func(req *http.Request, via []*http.Request) error {
-			return http.ErrUseLastResponse
-		},
-	}
-	//client.Transport = tr
-	req, err := http.NewRequest("POST", "https://academy.hackthebox.com/login", strings.NewReader(payload))
+	client, err := newClient()
 	if err != nil {
 		die(err)
 	}
-	req.Header.Add("Cookie", auth.cookies)
-	req.Header.Set("User-Agent", "Mozilla/5.0 (X11; Linux x86_64; rv:102.0) Gecko/20100101 Firefox/102.0")
-	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
 
-	resp, err := client.Do(req)
+	// Get academy cookies cookies
+	resp, err := client.Get("https://academy.hackthebox.com/login")
+	if err != nil {
+		die(err)
+	}
+
+	// Head over to the SSO login
+	resp, err = client.Get("https://academy.hackthebox.com/sso/redirect")
+	if err != nil {
+		die(err)
+	}
+
+	// Get CSRF-cookie for logging in
+	resp, err = client.Get("https://account.hackthebox.com/api/v1/csrf-cookie")
+	if err != nil {
+		die(err)
+	}
+
+	// Login to HackTheBox
+	credentials := Credentials{
+		Email:    email,
+		Password: password,
+	}
+	jsonData, err := json.Marshal(credentials)
+	if err != nil {
+		die(err)
+	}
+	req, err := http.NewRequest("POST", "https://account.hackthebox.com/api/v1/auth/login", bytes.NewBuffer(jsonData))
+	if err != nil {
+		die(err)
+	}
+	xsrfToken := getXSRFToken(client, "https://account.hackthebox.com/")
+	// Set headers
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("Accept-Language", "en-US,en;q=0.5")
+	req.Header.Set("Accept-Encoding", "identity")
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("X-Requested-With", "XMLHttpRequest")
+	req.Header.Set("X-Xsrf-Token", xsrfToken)
+	req.Header.Set("Origin", "https://account.hackthebox.com")
+	req.Header.Set("Referer", "https://account.hackthebox.com/login")
+	// Perform the request
+	resp, err = client.Do(req)
+	if err != nil {
+		die(err)
+	}
+
+	body, _ := io.ReadAll(resp.Body)
+
+	if resp.StatusCode != 200 {
+		fmt.Println("Authenticating to HackTheBox failed.")
+		fmt.Printf("Unexpected status code: %d\n", resp.StatusCode)
+		fmt.Println("Response body:", string(body))
+		os.Exit(1)
+	}
+	var loginResp LoginResponse
+	if err := json.Unmarshal(body, &loginResp); err != nil {
+		die(err)
+	}
+
+	// Make a request to the intended route
+	resp, err = client.Get(loginResp.IntendedRoute)
 	if err != nil {
 		die(err)
 	}
 	cookies := resp.Cookies()
 
-	return Auth{
-		loginToken: auth.loginToken,
-		cookies:    cookies[0].Name + "=" + cookies[0].Value + "; " + cookies[1].Name + "=" + cookies[1].Value,
-	}
+	return Auth{cookies: cookies[0].Name + "=" + cookies[0].Value + "; " + cookies[1].Name + "=" + cookies[1].Value}
 }
 
-func getLoginTokenAndCookies() Auth {
-	xsrfToken := ""
-	sessionToken := ""
-	resp, err := http.Get("https://academy.hackthebox.com/login")
-	if err != nil {
-		die(err)
+func (ua *userAgentTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if req.Header.Get("User-Agent") == "" {
+		req.Header.Set("User-Agent", ua.UserAgent)
 	}
-	cookies := resp.Cookies()
+	return ua.Transport.RoundTrip(req)
+}
+
+func newClient() (*http.Client, error) {
+	// For proxy debugging
+	//proxy, _ := url.Parse("http://localhost:8080")
+	//transport := &userAgentTransport{
+	//	Transport: &http.Transport{
+	//		Proxy:           http.ProxyURL(proxy),
+	//		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	//	},
+	//	UserAgent: "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0",
+	//}
+	transport := &userAgentTransport{
+		Transport: http.DefaultTransport,
+		UserAgent: "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0",
+	}
+	jar, err := cookiejar.New(&cookiejar.Options{PublicSuffixList: publicsuffix.List})
+	if err != nil {
+		return nil, err
+	}
+	return &http.Client{
+		Jar:       jar,
+		Transport: transport,
+	}, nil
+}
+
+func getXSRFToken(client *http.Client, urlStr string) string {
+	u, _ := url.Parse(urlStr)
+	cookies := client.Jar.Cookies(u)
 	for _, cookie := range cookies {
 		if cookie.Name == "XSRF-TOKEN" {
-			xsrfToken = cookie.Name + "=" + cookie.Value
-		} else if cookie.Name == "htb_academy_session" {
-			sessionToken = cookie.Name + "=" + cookie.Value
+			rawToken, err := url.QueryUnescape(cookie.Value)
+			if err != nil {
+				die(err)
+			}
+			return rawToken
 		}
 	}
-	if xsrfToken == "" || sessionToken == "" {
-		fmt.Println("WARNING: Required cookies not found.")
-	}
-	body, err := io.ReadAll(resp.Body)
-	if err != nil {
-		die(err)
-	}
-	content := string(body)
-	token := parseLoginToken(content)
-	return Auth{
-		loginToken: token,
-		cookies:    xsrfToken + ";" + sessionToken,
-	}
-}
-
-func parseLoginToken(htmlText string) string {
-	var token string
-	tkn := html.NewTokenizer(strings.NewReader(htmlText))
-
-	for {
-		tt := tkn.Next()
-		switch {
-
-		case tt == html.ErrorToken:
-			os.Exit(1)
-
-		case tt == html.StartTagToken:
-			t := tkn.Token()
-			if t.Data == "input" && t.Attr[0].Val == "hidden" {
-				token = t.Attr[2].Val
-				return token
-			}
-		case tt == html.EndTagToken:
-			t := tkn.Token()
-			if t.Data == "html" {
-				fmt.Println("Could not find token on login page, exiting.")
-				os.Exit(1)
-			}
-		}
-	}
-}
-
-func isLoggedIn(html string) bool {
-	if strings.Contains(html, "Sign in to continue to HTB Academy") {
-		return false
-	}
-	return true
+	return ""
 }
 
 func getModule(moduleUrl string, creds Auth) (string, []string) {
-	//proxy, _ := url.Parse("http://localhost:8080")
-	//tr := &http.Transport{Proxy: http.ProxyURL(proxy), TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}
 	client := &http.Client{}
-	//client.Transport = tr
 	req, err := http.NewRequest("GET", moduleUrl, nil)
 	if err != nil {
 		die(err)
@@ -133,11 +173,6 @@ func getModule(moduleUrl string, creds Auth) (string, []string) {
 		die(err)
 	}
 	content := string(body)
-	if !isLoggedIn(content) {
-		fmt.Println("Authentication failed. Please check that your credentials are correct.\n" +
-			"If your password has special characters such as quotes, ensure they are being escaped correctly.")
-		os.Exit(1)
-	}
 	moduleTitle := getModuleTitle(content)
 	pageUrls := getModulePages(content, moduleUrl)
 

--- a/src/export.go
+++ b/src/export.go
@@ -20,10 +20,6 @@ type userAgentTransport struct {
 	UserAgent string
 }
 
-type Auth struct {
-	cookies string
-}
-
 type Credentials struct {
 	Email    string `json:"email"`
 	Password string `json:"password"`
@@ -33,7 +29,7 @@ type LoginResponse struct {
 	IntendedRoute string `json:"intended_route"`
 }
 
-func authenticate(email, password string) Auth {
+func authenticate(email, password string) *http.Client {
 	client, err := newClient()
 	if err != nil {
 		die(err)
@@ -104,9 +100,8 @@ func authenticate(email, password string) Auth {
 	if err != nil {
 		die(err)
 	}
-	cookies := resp.Cookies()
 
-	return Auth{cookies: cookies[0].Name + "=" + cookies[0].Value + "; " + cookies[1].Name + "=" + cookies[1].Value}
+	return client
 }
 
 func (ua *userAgentTransport) RoundTrip(req *http.Request) (*http.Response, error) {
@@ -155,16 +150,8 @@ func getXSRFToken(client *http.Client, urlStr string) string {
 	return ""
 }
 
-func getModule(moduleUrl string, creds Auth) (string, []string) {
-	client := &http.Client{}
-	req, err := http.NewRequest("GET", moduleUrl, nil)
-	if err != nil {
-		die(err)
-	}
-	req.Header.Set("User-Agent", "Mozilla/5.0 (X11; Linux x86_64; rv:102.0) Gecko/20100101 Firefox/102.0")
-	req.Header.Add("Cookie", creds.cookies)
-
-	resp, err := client.Do(req)
+func getModule(moduleUrl string, client *http.Client) (string, []string) {
+	resp, err := client.Get(moduleUrl)
 	if err != nil {
 		die(err)
 	}
@@ -178,7 +165,7 @@ func getModule(moduleUrl string, creds Auth) (string, []string) {
 
 	var pagesContent []string
 	for _, pageUrl := range pageUrls {
-		pagesContent = append(pagesContent, extractPageContent(pageUrl, creds))
+		pagesContent = append(pagesContent, extractPageContent(pageUrl, client))
 	}
 
 	return moduleTitle, pagesContent
@@ -252,16 +239,8 @@ func getModulePages(htmlText string, moduleUrl string) []string {
 	return modulePages[1:]
 }
 
-func getModulePageContent(pageUrl string, creds Auth) string {
-	client := &http.Client{}
-	req, err := http.NewRequest("GET", pageUrl, nil)
-	if err != nil {
-		die(err)
-	}
-	req.Header.Add("Cookie", creds.cookies)
-	req.Header.Set("User-Agent", "Mozilla/5.0 (X11; Linux x86_64; rv:102.0) Gecko/20100101 Firefox/102.0")
-
-	resp, err := client.Do(req)
+func getModulePageContent(pageUrl string, client *http.Client) string {
+	resp, err := client.Get(pageUrl)
 	if err != nil {
 		die(err)
 	}
@@ -274,9 +253,9 @@ func getModulePageContent(pageUrl string, creds Auth) string {
 	return content
 }
 
-func extractPageContent(pageUrl string, creds Auth) string {
+func extractPageContent(pageUrl string, client *http.Client) string {
 	var result string
-	pageContent := getModulePageContent(pageUrl, creds)
+	pageContent := getModulePageContent(pageUrl, client)
 
 	doc, err := html.Parse(strings.NewReader(pageContent))
 	if err != nil {

--- a/src/main.go
+++ b/src/main.go
@@ -1,6 +1,7 @@
 package main
 
 import (
+	"fmt"
 	md "github.com/JohannesKaufmann/html-to-markdown"
 	"github.com/JohannesKaufmann/html-to-markdown/plugin"
 	"os"
@@ -9,9 +10,12 @@ import (
 
 func main() {
 	options := getArguments()
+	fmt.Println("Authenticating with HackTheBox...")
 	session := authenticate(options.email, options.password)
+	fmt.Println("Downloading requested module...")
 	title, content := getModule(options.moduleUrl, session)
 	if options.localImages {
+		fmt.Println("Downloading module images...")
 		content = getImagesLocally(content)
 	}
 
@@ -21,6 +25,7 @@ func main() {
 	if err != nil {
 		die(err)
 	}
+	fmt.Println("Finished downloading module!")
 }
 
 func htmlToMarkdown(html []string) string {

